### PR TITLE
Introduce new `cardano-cli latest transaction build-estimate` command

### PR DESCRIPTION
**Pull request by Jimbo4350**
# Changelog

```yaml
- description: |
    Introduce new `cardano-cli latest transaction build-estimate` command which will 
    produce a balanced transaction body without requiring a connection to a live node

  type:
  - feature  
```

# Context

Additional context for the PR goes here. If the PR fixes a particular issue please provide a [link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=) to the issue.

# How to trust this PR

Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

**Comment by Jimbo4350** [comment]
```bash
cardano-cli latest transaction build-estimate                                                                 
Usage: cardano-cli latest transaction build-estimate 
                                                       [ --script-valid
                                                       | --script-invalid
                                                       ]
                                                       --shelley-key-witnesses INT
                                                       [--byron-key-witnesses Int]
                                                       --protocol-params-file FILE
                                                       --total-utxo-value VALUE
                                                       (--tx-in TX-IN
                                                         [ --spending-tx-in-reference TX-IN
                                                           ( --spending-plutus-script-v2
                                                           | --spending-plutus-script-v3
                                                           )
                                                           ( --spending-reference-tx-in-datum-cbor-file CBOR FILE
                                                           | --spending-reference-tx-in-datum-file JSON FILE
                                                           | --spending-reference-tx-in-datum-value JSON VALUE
                                                           | --spending-reference-tx-in-inline-datum-present
                                                           )
                                                           ( --spending-reference-tx-in-redeemer-cbor-file CBOR FILE
                                                           | --spending-reference-tx-in-redeemer-file JSON FILE
                                                           | --spending-reference-tx-in-redeemer-value JSON VALUE
                                                           )
                                                           --spending-reference-tx-in-execution-units (INT, INT)
                                                         | --simple-script-tx-in-reference TX-IN
                                                         | --tx-in-script-file FILE
                                                           [
                                                             ( --tx-in-datum-cbor-file CBOR FILE
                                                             | --tx-in-datum-file JSON FILE
                                                             | --tx-in-datum-value JSON VALUE
                                                             | --tx-in-inline-datum-present
                                                             )
                                                             ( --tx-in-redeemer-cbor-file CBOR FILE
                                                             | --tx-in-redeemer-file JSON FILE
                                                             | --tx-in-redeemer-value JSON VALUE
                                                             )
                                                             --tx-in-execution-units (INT, INT)]
                                                         ])
                                                       [--read-only-tx-in-reference TX-IN]
                                                       [ --required-signer FILE
                                                       | --required-signer-hash HASH
                                                       ]
                                                       [--tx-in-collateral TX-IN]
                                                       [--tx-out-return-collateral ADDRESS VALUE]
                                                       [--tx-out ADDRESS VALUE
                                                         [ --tx-out-datum-hash HASH
                                                         | --tx-out-datum-hash-cbor-file CBOR FILE
                                                         | --tx-out-datum-hash-file JSON FILE
                                                         | --tx-out-datum-hash-value JSON VALUE
                                                         | --tx-out-datum-embed-cbor-file CBOR FILE
                                                         | --tx-out-datum-embed-file JSON FILE
                                                         | --tx-out-datum-embed-value JSON VALUE
                                                         | --tx-out-inline-datum-cbor-file CBOR FILE
                                                         | --tx-out-inline-datum-file JSON FILE
                                                         | --tx-out-inline-datum-value JSON VALUE
                                                         ]
                                                         [--tx-out-reference-script-file FILE]]
                                                       --change-address ADDRESS
                                                       [--mint VALUE
                                                         ( --mint-script-file FILE
                                                           [
                                                             ( --mint-redeemer-cbor-file CBOR FILE
                                                             | --mint-redeemer-file JSON FILE
                                                             | --mint-redeemer-value JSON VALUE
                                                             )
                                                             --mint-execution-units (INT, INT)]
                                                         | --simple-minting-script-tx-in-reference TX-IN
                                                           --policy-id HASH
                                                         | --mint-tx-in-reference TX-IN
                                                           ( --mint-plutus-script-v2
                                                           | --mint-plutus-script-v3
                                                           )
                                                           ( --mint-reference-tx-in-redeemer-cbor-file CBOR FILE
                                                           | --mint-reference-tx-in-redeemer-file JSON FILE
                                                           | --mint-reference-tx-in-redeemer-value JSON VALUE
                                                           )
                                                           --mint-reference-tx-in-execution-units (INT, INT)
                                                           --policy-id HASH
                                                         )]
                                                       [--invalid-before SLOT]
                                                       [--invalid-hereafter SLOT]
                                                       [--certificate-file FILE
                                                         [ --certificate-script-file FILE
                                                           [
                                                             ( --certificate-redeemer-cbor-file CBOR FILE
                                                             | --certificate-redeemer-file JSON FILE
                                                             | --certificate-redeemer-value JSON VALUE
                                                             )
                                                             --certificate-execution-units (INT, INT)]
                                                         | --certificate-tx-in-reference TX-IN
                                                           ( --certificate-plutus-script-v2
                                                           | --certificate-plutus-script-v3
                                                           )
                                                           ( --certificate-reference-tx-in-redeemer-cbor-file CBOR FILE
                                                           | --certificate-reference-tx-in-redeemer-file JSON FILE
                                                           | --certificate-reference-tx-in-redeemer-value JSON VALUE
                                                           )
                                                           --certificate-reference-tx-in-execution-units (INT, INT)
                                                         ]]
                                                       [--withdrawal WITHDRAWAL
                                                         [ --withdrawal-script-file FILE
                                                           [
                                                             ( --withdrawal-redeemer-cbor-file CBOR FILE
                                                             | --withdrawal-redeemer-file JSON FILE
                                                             | --withdrawal-redeemer-value JSON VALUE
                                                             )
                                                             --withdrawal-execution-units (INT, INT)]
                                                         | --withdrawal-tx-in-reference TX-IN
                                                           ( --withdrawal-plutus-script-v2
                                                           | --withdrawal-plutus-script-v3
                                                           )
                                                           ( --withdrawal-reference-tx-in-redeemer-cbor-file CBOR FILE
                                                           | --withdrawal-reference-tx-in-redeemer-file JSON FILE
                                                           | --withdrawal-reference-tx-in-redeemer-value JSON VALUE
                                                           )
                                                           --withdrawal-reference-tx-in-execution-units (INT, INT)
                                                         ]]
                                                       [--tx-total-collateral INTEGER]
                                                       [--reference-script-size NATURAL]
                                                       [ --json-metadata-no-schema
                                                       | --json-metadata-detailed-schema
                                                       ]
                                                       [--auxiliary-script-file FILE]
                                                       [ --metadata-json-file FILE
                                                       | --metadata-cbor-file FILE
                                                       ]
                                                       [--update-proposal-file FILE]
                                                       --out-file FILE

  Build a balanced transaction without access to a live node (automatically estimates fees)
```

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs [hunk 1]
```diff
@@ -9,8 +9,10 @@
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TupleSections #-}
 {-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
 
 {-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+{-# OPTIONS_GHC -fwarn-incomplete-patterns #-}
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs]: Why is this needed?

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs [hunk 2]
```diff
@@ -245,6 +248,213 @@ runTransactionBuildCmd
...
+      readTxUpdateProposal w updateProposalFile & firstExceptT TxCmdReadTextViewFileError
+    _ -> pure TxUpdateProposalNone
+
+  requiredSigners  <- mapM (firstExceptT TxCmdRequiredSignerError .  newExceptT . readRequiredSigner) reqSigners
+
+  mReturnCollateral <- case mReturnColl of
+                         Nothing -> return Nothing
+                         Just retColl -> do
+                          txOut <- toTxOutInShelleyBasedEra sbe retColl
+                          return $ Just txOut
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs]:
```suggestion
  mReturnCollateral <- mapM (toTxOutInShelleyBasedEra sbe) mReturnColl
```

File: cardano-cli/test/cardano-cli-golden/files/golden/help/conway_transaction.cli [hunk 3]
```diff
@@ -24,6 +25,9 @@ Available commands:
                            Please note the order of some cmd options is crucial. If used incorrectly may produce undesired tx body. See nested [] notation above for details.[0;22;23;24m
   build                    Build a balanced transaction (automatically calculates fees)
 
+                           Please note the order[93;22;23;24m of some cmd options is crucial. If used incorrectly may produce undesired tx body. See nested [] notation above for details.[0;22;23;24m
```
Comment (carbolymer) [inline cardano-cli/test/cardano-cli-golden/files/golden/help/conway_transaction.cli:28]: where those escape characters come from?

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs (hunk 1 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs]: It shouldn't be. I was trying to understand why GHC did not complain about: https://github.com/IntersectMBO/cardano-cli/pull/728/files#diff-a38343c1825e6e334224d0a8a1ba78a458652970f00c54f5fb0d040be28f822eR787

Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs]: :smiling_face_with_tear: 
```
              ┃   │ ━━━━ stderr ━━━━
                ┃   │ 
                ┃   │ cardano-cli: user error (Pattern match failure in 'do' block at src/Cardano/CLI/EraBased/Run/Transaction.hs:787:3-26)
```

File: cardano-cli/test/cardano-cli-golden/files/golden/help/conway_transaction.cli (hunk 3 above)
Comment (Jimbo4350) [reply cardano-cli/test/cardano-cli-golden/files/golden/help/conway_transaction.cli:28]: `opt-parse-applicative`. It's the colouring of the text.

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs (hunk 1 above)
Comment (carbolymer) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs]: But there's no pattern matching in that fragment:
```haskell
  validatedTxCerts
     <- hoistEither
          . first TxCmdTxCertificatesValidationError
          $ validateTxCertificates era certsAndMaybeScriptWits
```
strange

**Review by carbolymer — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs (hunk 1 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs]: I updated it.
